rfcs/0016-better-logging.md

```diff
@@ -0,0 +1,153 @@
+  "Error adding breadcrumb $breadcrumb to transition frontier: %s"
+  (Error.to_string err)
+```
+
+which would produce a log like (where the `<>`'s are replaced)
+
+```json
+{
+  "timestamp": <now>,
+  "level": "error",
+  "source": {
+    "module": "<__MODULE__>",
+    "location": "<__loc__>"
+  },
+  "message": "Error adding breadcrumb $breadcrumb to transition frontier: <err>",
+  "metadata": {
+    "breadcrumb": {<breadcrumb...>},
+    <context...>
+  }
+}
+```
+
+and would be formatted as
+
+```
+Error adding breadcrumb {...} to transition frontier: <err>
+```
+
+### Log Processor
+[detailed-design-log-processor]: #detailed-design-log-processor
+
+While the logging format is still in json, meaning that programs such as `jq` et al. can be used to process logs, this has been too slow in practice. Specifically, in the case of `jq`, there is limited support for handling files full of multiple json objects (statically or streaming), so `jq` needs to be reinvoked for every line in the log, which involves re-parsing and processing filter expressions for each log entry. Instead, we will go back to writing our own log processor utility. This makes additional sense in the context that we want to perform and configure interpolation for our log messages.
+
+The log processor should support some basic configuration to control how interpolation of the message string is performed. For instance, the default could be to attempt to keep all message under a specific length (say `< 80`) and to elide interpolation otherwise. Another option could be to constrain a max length of the interpolated value, or even constrain interpolation by the type of the value (for instance, elide all objects and interpolate other values). And yet another would be to just always interpolate the entire values, or print messages in a raw format with the relevant metadata displayed below the message. These various options should be added as they are needed by developers, so I will not lay out the specific requirements here.
+
+The log processor should support a simple `jq`-esque filter language. The scope of this is *significantly* smaller than that of `jq` as it only needs the ability to form simple predicates on a json object. An example of what this language could look like is speced out below in BNF. It mostly follows javascript syntax with little divergence, except that it is much more constrained in scope. The toplevel rule is `<bool_exp>` since this is a filter language.
+
+```bnf
+<bool> ::= "true" | "false"
+
+<digit> ::= "0".."9"
+<integer> ::= <digit> | <integer> <digit>
+
+<alpha> ::= "a".."z" | "A".."Z" | "_"
+<char> ::= <alpha> | "0" .. "9"
+<escape> ::= '"' | '\' | '/' | 'b' | 'n' | 'r' | 't'
+<text> ::= "" | '\' <escape> <text> | <char> <text>
+<string> ::= '"' <text> '"'
+
+<ident> ::= <alpha> <text>
+
+<literal> ::= <bool> | <integer> | <string>
+<access_exp> ::= "." <ident> | "[" <string> "]" | "[" <integer> "]"
```

Comment:
recursive `.` accesses would be nice

Reply:
This was added in the implementation branch. I will update the bnf grammar to match the additions.